8246037: Shenandoah: update man pages to mention -XX:+UseShenandoahGC

rgithubli opened the pull request:
Add documentation of Shenandoah to java man page

Aside from `-XX:+UseShenandoahGC`, I picked flags from [shenandoah_globals.hpp](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp) that are at product level but not experimental / diagnostic to avoid overwhelming info. Two additional flags match: `ShenandoahGCMode` and `ShenandoahGCHeuristics`

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8246037](https://bugs.openjdk.org/browse/JDK-8246037): Shenandoah: update man pages to mention -XX:+UseShenandoahGC (**Enhancement** - P4)


### Reviewers
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - **Reviewer**) Review applies to [cbf1819a](https://git.openjdk.org/jdk/pull/26907/files/cbf1819a84ef6349b18198cc4eeacaca6cc9e449)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - **Reviewer**)
 * [Cesar Soares Lucas](https://openjdk.org/census#cslucas) (@JohnTortugo - Committer) Review applies to [1c8d334a](https://git.openjdk.org/jdk/pull/26907/files/1c8d334a6b80b008a9c3661f78d6b1f89f939099)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/26907/head:pull/26907` \
`$ git checkout pull/26907`

Update a local copy of the PR: \
`$ git checkout pull/26907` \
`$ git pull https://git.openjdk.org/jdk.git pull/26907/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 26907`

View PR using the GUI difftool: \
`$ git pr show -t 26907`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/26907.diff">https://git.openjdk.org/jdk/pull/26907.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/26907#issuecomment-3215280393)
</details>

shipilev commented on src/java.base/share/man/java.md:
```diff
@@ -2883,6 +2883,48 @@ Java HotSpot VM.
+    `satb`
+    :   Snapshot-at-the-beginning concurrent GC (three pass mark-evac-update).
+
+    `passive`
+    :   Stop the world GC only (either degenerated or full). This mode is
+        diagnostic, and must be enabled via `-XX:+UnlockDiagnosticVMOptions`.
```
I don't think we add diagnostic options to the generic options list. These are not for end users, and we do not surface them unnecessarily.

rgithubli replied:
Will remove

shipilev commented on src/java.base/share/man/java.md:
```diff
@@ -2883,6 +2883,48 @@ Java HotSpot VM.
+    to `satb`. Among other things, this defines which barriers are in use.
+    Possible mode values include the following:
+
+    `satb`
+    :   Snapshot-at-the-beginning concurrent GC (three pass mark-evac-update).
```
This is a long-standing problem with the name. It kinda implies `generational` is not `satb`. It actually is. Maybe we should rename `satb` to `single` or something. A lot of bikeshedding would ensue.

ysramakrishna replied:
I agree. For non-experimental options, we can follow the established deprecate & rename protocol.

May be we could, in the inerim, add verbage in the description to avoid the confusion in the current naming as described by Aleksey.

shipilev commented on src/java.base/share/man/java.md:
```diff
@@ -2883,6 +2883,48 @@ Java HotSpot VM.
+
+    `aggressive`
+    :   Run GC continuously, try to evacuate everything. This heuristics is
+        diagnostic, and must be enabled via -XX:+UnlockDiagnosticVMOptions.
```
Same here: diagnostic option.

shipilev commented on src/java.base/share/man/java.md:
```diff
@@ -2883,6 +2883,48 @@ Java HotSpot VM.
+    the GC cycle.
+
+    `static`
+    :   Trigger GC when free heap falls below the ShenandoahMinFreeThreshold.
```
Same here: `ShenandoahMinFreeThreshold` is experimental, it is not worth mentioning here. In fact, there are tunables for `adaptive` and `compact` as well, yet we do not (correctly!) mention them.

rgithubli replied:
True, other options respond to `ShenandoahMinFreeThreshold` as well. It's kinda hard to describe `static` option without mentioning the threshold. 
- We could make the doc vague as in the [shenandoah_globals.hpp](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp#L147): `trigger GC when free heap falls below the threshold`. 
- I now wonder why `static` is a non experimental flag - it has to be used with other experimental flags. Maybe just drop static from man page as well?

ysramakrishna replied:
`below a specified threshold` might sound appropriately vague and render the flag unusable unless one consulted the code, underlining the experimental nature of it.

If the collector respects `static` as specified, it can be considered a good guardrail to use in anger. In fact, I know of teams that do in production. (For that matter, there are a couple who use `compact` as well.)

However, I am fine with eliding all experimental options, and in turn any non-experimental options that depend on experimental options (which however seems to be an anti-pattern to be avoided, if possible).

rgithubli commented:
Updated for other comments except for renaming `satb`: instead, I added additional explanation to say satb is a single generation GC, as well as generational shenandoah is also an satb gc.

ysramakrishna commented on src/java.base/share/man/java.md:
```diff
@@ -2883,6 +2883,43 @@ Java HotSpot VM.
+
+`-XX:ShenandoahGCHeuristics=`*heuristics*
+:   Sets the heuristics for Shenandoah GC to use. By default, this option is
+    set to `adaptive`. This fine-tunes the GC mode selected, by choosing when
+    to start the GC, how much to process on each cycle, and what other features
+    to automatically enable. Possible heuristics are following:
+
+    `adaptive`
+    :   To maintain the given amount of free heap at all times, even during
+        the GC cycle.
```
May be this part (and in `shenandoah_globals.hpp`) could be changed so as not to repeat the description of `adaptive`, by stating that the default setting is adaptive, but replacing the current description of `adaptive` at line 2918 with the text earlier at line 2909-2911.

ysramakrishna reviewed (approved):
One small comment/suggestion. Rest looks good to me.

Approving modulo that suggestion.

ysramakrishna commented on src/java.base/share/man/java.md:
```diff
@@ -2883,6 +2883,43 @@ Java HotSpot VM.
+        generational. Please see [JEP 404](https://openjdk.org/jeps/404) and
+        [JEP 521](https://openjdk.org/jeps/521) for its advantages and risks.
+
+`-XX:ShenandoahGCHeuristics=`*heuristics*
```
Good point.

rgithubli commented:
/integrate

ysramakrishna commented on src/java.base/share/man/java.md:
```diff
@@ -2908,7 +2908,10 @@ Java HotSpot VM.
-    to automatically enable. Possible heuristics are following:
+    to automatically enable.
+
+    Possible heuristics are the following (when `-XX:ShenandoahGCMode` is
+    `generational`, the only supported option is the default, `adaptive`):
```
Move the parenthetic part to the previous paragraph and out of parentheses.

ysramakrishna reviewed (approved):
Approved modulo a comment.

I also think we should to the extent possible make the listing of text common to man page and options list consistent. An example is the recent improvement you made in your most recent update to the man page for the heuristics option.  There may be other changes to the man page that came about as a result of this review. It would make sense to make those changes, where applicable, also to shenandoah_globals. This latter need not hold up this ticket but can be addressed separately in a follow up.

JohnTortugo reviewed (approved):
Ship it!

rgithubli commented:
/integrate

kdnilsen commented:
/sponsor
